File: traversal/pre-order/recursion.ts
import { BinaryTreeNode } from 'data-structures/binary-tree'

/**
 * @leetcode https://leetcode-cn.com/problems/binary-tree-preorder-traversal/
 */
export function preOrderTraversal<V>(node: BinaryTreeNode<V> | null) {
  const answer: V[] = []
  traversal(node, answer)
  return answer
}

function traversal<V>(node: BinaryTreeNode<V> | null, answer: V[]) {
  if (node === null) return

  answer.push(node.value)
  traversal(node.left, answer)
  traversal(node.right, answer)
}

File: traversal/pre-order/recursion.go
package preorder

import binarytree "algorithms/data-structures/src/binary_tree"

// Recursive for binary tree pre-order traversal
func Recursive(root *binarytree.Node) []int {
	answer := []int{}
	traversal(root, &answer)
	return answer
}

func traversal(root *binarytree.Node, answer *[]int) {
	if root == nil {
		return
	}

	*answer = append(*answer, root.Val.(int))
	traversal(root.Left, answer)
	traversal(root.Right, answer)
}
